flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Expressions.java

```diff
@@ -0,0 +1,527 @@
+	 * based on a evaluated boolean condition.
+	 *
+	 * <p>e.g. ifThenElse(42 > 5, "A", "B") leads to "A"
+	 *
+	 * @param condition boolean condition
+	 * @param ifTrue expression to be evaluated if condition holds
+	 * @param ifFalse expression to be evaluated if condition does not hold
+	 */
+	public static ApiExpression ifThenElse(Object condition, Object ifTrue, Object ifFalse) {
+		return apiCall(
+			BuiltInFunctionDefinitions.IF,
+			objectToExpression(condition),
+			objectToExpression(ifTrue),
+			objectToExpression(ifFalse));
+	}
+
+	/**
+	 * Creates an expression that selects a range of columns. It can be used wherever an array of
+	 * expression is accepted such as function calls, projections, or groupings.
+	 *
+	 * <p>A range can either be index-based or name-based. Indices start at 1 and boundaries are
+	 * inclusive.
+	 *
+	 * <p>e.g. withColumns('b to 'c) or withColumns('*)
+	 */
+	public static ApiExpression withColumns(Object head, Object... tail) {
+		return apiCall(
+			BuiltInFunctionDefinitions.WITH_COLUMNS,
+			Stream.concat(
+				Stream.of(head),
+				Stream.of(tail)
+			).map(ApiExpressionUtils::objectToExpression)
+				.toArray(Expression[]::new));
+	}
+
+	/**
+	 * Creates an expression that selects all columns except for the given range of columns. It can
+	 * be used wherever an array of expression is accepted such as function calls, projections, or
+	 * groupings.
+	 *
+	 * <p>A range can either be index-based or name-based. Indices start at 1 and boundaries are
+	 * inclusive.
+	 *
+	 * <p>e.g. withoutColumns('b to 'c) or withoutColumns('c)
+	 */
+	public static ApiExpression withoutColumns(Object head, Object... tail) {
+		return apiCall(
+			BuiltInFunctionDefinitions.WITHOUT_COLUMNS,
+			Stream.concat(
+				Stream.of(head),
+				Stream.of(tail)
+			).map(ApiExpressionUtils::objectToExpression)
+				.toArray(Expression[]::new));
+	}
+
+	/**
+	 * A call to a function that will be looked up in a catalog.
+	 */
+	public static ApiExpression call(String functionName, Object... params) {
+		return new ApiExpression(ApiExpressionUtils.lookupCall(
+			functionName,
+			Arrays.stream(params).map(ApiExpressionUtils::objectToExpression).toArray(Expression[]::new)));
+	}
+
+	/**
+	 * A call to an inline function. For functions registered in a catalog use {@link #call(String, Object...)}.
+	 */
+	public static ApiExpression call(FunctionDefinition scalarFunction, Object... params) {
+		return apiCall(
+			scalarFunction,
+			Arrays.stream(params).map(ApiExpressionUtils::objectToExpression).toArray(Expression[]::new));
+	}
+
+	private static ApiExpression apiCall(FunctionDefinition functionDefinition, Expression... args) {
+		return new ApiExpression(new UnresolvedCallExpression(functionDefinition, Arrays.asList(args)));
+	}
+
+	/**
+	 * Java API class that gives access to expressions operations.
+	 */
+	public static final class ApiExpression extends BaseExpressions<Object, ApiExpression> implements Expression {
```

Comment:
Unfortunately this will not work. The Java's Expression DSL return the `ApiExpression` e.g.
```
class Expressions {
   ApiExpression $(String name);
}
```
thus the `ApiExpression` must be public so that users can access methods defined in `BaseExpression`. We could return `BaseExpressions` instead, but I don't think it would be any better.

Moreover I think the original idea was that if we have some non-static Java specific methods we can put them in `ApiExpression`.

Reply:
Sure, ApiExpression must be public. I was just suggesting to have a hidden constructor and a top-level class.

Reply:
Ah, misread that, sorry. Will do